File: package/main/src/String/trimEndCharacters.ts
/**
 * Removes specified characters from the end of a string
 *
 * @param {string} string_ - The input string to trim
 * @param {string} chars - Characters to remove from the end
 * @returns {string} A new string with specified characters removed from the end
 * @example
 * ```typescript
 * trimEndCharacters("hello!!!", "!"); // Returns: "hello"
 * trimEndCharacters("123---", "-");   // Returns: "123"
 * trimEndCharacters("abc123", "xyz"); // Returns: "abc123"
 * ```
 */
export const trimEndCharacters = (string_: string, chars: string): string => {
  // Use Set for O(1) character lookups instead of O(m) string.includes()
  const charSet = new Set(chars);
  let endIndex = string_.length - 1;
  while (endIndex >= 0 && charSet.has(string_[endIndex])) {
    endIndex--;
  }
  return string_.slice(0, endIndex + 1);
};

File: package/umt_python/src/string/trim_end_characters.py
def trim_end_characters(string_: str, chars: str) -> str:
    """
    Removes specified characters from the end of a string.

    Args:
        string_: The input string to trim.
        chars: Characters to remove from the end.

    Returns:
        A new string with specified characters removed from the end.

    Example:
        >>> trim_end_characters("hello!!!", "!")
        'hello'
        >>> trim_end_characters("123---", "-")
        '123'
        >>> trim_end_characters("abc123", "xyz")
        'abc123'
    """
    end_index = len(string_)
    while end_index > 0 and string_[end_index - 1] in chars:
        end_index -= 1
    return string_[:end_index]
